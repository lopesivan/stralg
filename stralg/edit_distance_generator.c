
#include "edit_distance_generator.h"
#include "cigar.h"

#include <string.h>
#include <stdio.h>

struct recursive_constant_data {
    const char *buffer_front;
    const char *cigar_front;
    const char *alphabet;
    char *simplify_cigar_buffer;
};

static void recursive_generator(const char *pattern, char *buffer, char *cigar,
                                int max_edit_distance,
                                struct recursive_constant_data *data,
                                edits_callback_func callback,
                                void *callback_data)
{
    if (*pattern == '\0') {
        // no more pattern to match ... terminate the buffer and call back
        *buffer = '\0';
        *cigar = '\0';
        simplify_cigar(data->cigar_front, data->simplify_cigar_buffer);
        callback(data->buffer_front, data->simplify_cigar_buffer, callback_data);
        
    } else if (max_edit_distance == 0) {
        // we can't edit any more, so just move pattern to buffer and call back
        size_t rest = strlen(pattern);
        for (size_t i = 0; i < rest; ++i) {
            buffer[i] = pattern[i];
            cigar[i] = '=';
        }
        buffer[rest] = cigar[rest] = '\0';
        simplify_cigar(data->cigar_front, data->simplify_cigar_buffer);
        callback(data->buffer_front, data->simplify_cigar_buffer, callback_data);
        
    } else {
        // --- time to recurse --------------------------------------
        // deletion
        *cigar = 'I';
        recursive_generator(pattern + 1, buffer, cigar + 1,
                            max_edit_distance - 1, data,
                            callback, callback_data);
        // insertion
        for (const char *a = data->alphabet; *a; a++) {
            *buffer = *a;
            *cigar = 'D';
            recursive_generator(pattern, buffer + 1, cigar + 1,
                                max_edit_distance - 1, data,
                                callback, callback_data);
        }
        // match / substitution
        for (const char *a = data->alphabet; *a; a++) {
            if (*a == *pattern) {
                *buffer = *a;
                *cigar = '=';
                recursive_generator(pattern + 1, buffer + 1, cigar + 1,
                                    max_edit_distance, data,
                                    callback, callback_data);
            } else {
                *buffer = *a;
                *cigar = 'X';
                recursive_generator(pattern + 1, buffer + 1, cigar + 1,
                                    max_edit_distance - 1, data,
                                    callback, callback_data);
            }
        }
    }
}

void generate_all_neighbours(const char *pattern,
                             const char *alphabet,
                             int max_edit_distance,
                             edits_callback_func callback,
                             void *callback_data)
{
    size_t n = strlen(pattern) + max_edit_distance + 1;
    char buffer[n];
    char cigar[n], cigar_buffer[n];
    struct recursive_constant_data data = { buffer, cigar, alphabet, cigar_buffer };
    recursive_generator(pattern, buffer, cigar, max_edit_distance, &data,
                        callback, callback_data);
}
