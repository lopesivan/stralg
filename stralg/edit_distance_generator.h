
#ifndef EDIT_DISTANCE_GENERATOR_H
#define EDIT_DISTANCE_GENERATOR_H

#include <stddef.h>
#include <stdbool.h>

typedef void (*edits_callback_func)(
    const char *string,
    const char *cigar,
    void * data
);

void generate_all_neighbours(
    const char *pattern,
    const char *alphabet,
    int max_edit_distance,
    edits_callback_func callback,
    void *callback_data
);

struct edit_iter_frame;
struct edit_iter {
    const char *pattern;
    const char *alphabet;

    char *buffer;
    char *cigar;
    char *simplify_cigar_buffer;

    struct edit_iter_frame *frames;
};
struct edit_pattern {
    const char *pattern;
    const char *cigar;
};

// Do not muck about with the pattern between
// creating an iterator and you deallocate it again.
// The iterator code will get angry if you do,
// and it will take it out on your program.
void edit_init_iter(
    struct edit_iter *iter,
    const char *pattern,
    const char *alphabet,
    int max_edit_distance
);

bool edit_next_pattern(
    struct edit_iter *iter,
    struct edit_pattern *result
);

// This function frees the resources stored in the
// iterator structor, not the struct itself. If it is
// heap allocated you are responsible for freeing
// the structure.
void edit_dealloc_iter(
    struct edit_iter *iter
);

#endif


