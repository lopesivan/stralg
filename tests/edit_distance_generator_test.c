
#include <edit_distance_generator.h>

#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdio.h>

void callback(const char *string, const char *cigar, void *data)
{
    printf("%s %s\n", string, cigar);
}

int main(int argc, char * argv[])
{
    const char *pattern = "acca";
    const char *alphabet = "acgt";

    struct edit_iter iter;
    struct edit_pattern result;
    edit_init_iter(&iter, pattern, alphabet, 1);
    while (edit_next_pattern(&iter, &result)) {
        printf("%s %s\n",
        result.pattern, result.cigar);
    }
    edit_dealloc_iter(&iter);

    return EXIT_SUCCESS;
}
