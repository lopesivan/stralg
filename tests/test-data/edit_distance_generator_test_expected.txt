aaca 4M
aacca 1D4M
aacca 1M1D3M
aca 1M1I2M
aca 2M1I1M
acaa 4M
acaca 2M1D2M
acc 3M1I
acca 4M
accaa 3M1D1M
accc 4M
accca 1M1D3M
accca 2M1D2M
accca 3M1D1M
accg 4M
accga 3M1D1M
acct 4M
accta 3M1D1M
acga 4M
acgca 2M1D2M
acta 4M
actca 2M1D2M
agca 4M
agcca 1M1D3M
atca 4M
atcca 1M1D3M
cacca 1D4M
cca 1I3M
ccca 4M
gacca 1D4M
gcca 4M
tacca 1D4M
tcca 4M
